### PR TITLE
[SPARK-22945][SQL] add java UDF APIs in the functions object

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently Scala users can use UDF like
```
val foo = udf((i: Int) => Math.random() + i).asNondeterministic
df.select(foo('a))
```
Python users can also do it with similar APIs. However Java users can't do it, we should add Java UDF APIs in the functions object.

## How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 1]
```diff
@@ -110,29 +110,29 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
...
-            ScalaUDF(func, dataType, e, inputTypes.getOrElse(Nil), Some(name), nullable, udfDeterministic = true)
-          } else {
-             throw new AnalysisException("Invalid number of arguments for function " + name +
-               ". Expected: $x; Found: " + e.length)
-          }
-          functionRegistry.createOrReplaceTempFunction(name, builder)
-          val udf = UserDefinedFunction(func, dataType, inputTypes).withName(name)
-          if (nullable) udf else udf.asNonNullable()
-        }""")
+        |/**
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: just code style change, to use the multiline string format.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala [hunk 2]
```diff
@@ -66,6 +66,7 @@ case class UserDefinedFunction protected[sql] (
    *
    * @since 1.3.0
    */
+  @scala.annotation.varargs
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala]: to make it java friendly, or shall we create a new method for java?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -3254,42 +3254,66 @@ object functions {
...
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  // Scala UDF functions
+  //////////////////////////////////////////////////////////////////////////////////////////////
+
   /**
-   * Defines a deterministic user-defined function of 0 arguments as user-defined
-   * function (UDF). The data types are automatically inferred based on the function's
-   * signature. To change a UDF to nondeterministic, call the API
-   * `UserDefinedFunction.asNondeterministic()`.
+   * Defines a Scala closure of 0 arguments as user-defined function (UDF).
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: to make the document consistent with `UDFRegistration.register`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -3461,13 +3485,172 @@ object functions {
...
+   */
+  def udf(f: UDF0[_], returnType: DataType): UserDefinedFunction = {
+    val func = f.asInstanceOf[UDF0[Any]].call()
+    UserDefinedFunction(() => func, returnType, inputTypes = None)
+  }
+
+  /**
+   * Defines a Java UDF1 instance as user-defined function (UDF).
+   * The caller must specify the output data type, and there is no automatic input type coercion.
+   * By default the returned UDF is deterministic, to change it to nondeterministic, call the
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Nit: `, to` -> `. To`

**Comment by gatorsmile** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3

**Pull request closed without merging**